test/make_dist.py

```diff
@@ -52,115 +51,78 @@ def build_dist():
-                if os.isatty(sys.stderr.fileno()):
-                    total_size = 0
-                else:
-                    total_size = None
-                MB = 10**6
-                # read tar into a stringio, as the stream is not seekable and tar requires that
-                while True:
-                    block = response.read(MB)
-                    if len(block) == 0:
-                        break
-                    if total_size is not None:
-                        total_size += len(block)
-                        sys.stderr.write(f"\r{ total_size // MB } MB")
-
-                    tario.write(block)
-
-                # clear the download progress in tty mode
-                if total_size is not None:
-                    sys.stderr.write("\r                             \r")
-
-                break
-
-        except urllib.error.HTTPError as e:
+            subprocess.check_call(["git", "-C", dist_git_checkout, "fetch", "origin", "sha-" + sha, "--depth=1"])
+            break
+        except subprocess.CalledProcessError:
             retries -= 1
 
             if retries == 0:
-                message(f"make_dist: Downloading {download_url} failed:", e)
-                return None
+                message(f"make_dist: Downloading pre-built dist for SHA {sha} failed")
+                return False
 
-            message(f"make_dist: {download_url} not yet available, waiting...")
+            message(f"make_dist: pre-built dist for {sha} not yet available, waiting...")
             time.sleep(30)
 
-    tario.seek(0)
-    with tarfile.open(fileobj=tario) as ftar:
-        names = ftar.getnames()
-        try:
-            names.remove('.')
-        except ValueError:
-            pass
-        if len(names) != 1 or not names[0].endswith(".tar.xz"):
-            message("make_dist: expected tar with exactly one tar.xz member")
-            return None
-        ftar.extract(names[0])
-        tar_path = os.path.realpath(names[0])
-
-    # Extract npm/webpack related files locally for speeding up the build and allowing integration tests to run
-    unpack_paths = [d for d in ["dist", "node_modules", "package-lock.json", "tools/debian/copyright"] if not os.path.exists(d)]
-    if unpack_paths:
-        message("make_dist: Extracting from tarball:", ' '.join(unpack_paths))
-        prefix = os.path.basename(tar_path).split('.tar')[0] + '/'
-        prefixed_unpack_paths = [prefix + d for d in unpack_paths]
-        subprocess.check_call(["tar", "--touch", "--strip-components=1", "-xf", tar_path] + prefixed_unpack_paths)
-        # after the above, dist manifests may be newer than copyright due to tar touching order
-        # restore file time stamps as they would be after a build
-        subprocess.check_call(["touch", "tools/debian/copyright"])
-
-    return tar_path
+    # built files need to have current time, not commit time, to satisfy make dependencies
```

Comment:
I feel like with the order that we unpack things, we could probably *always* touch.  That would be a nice cleanup to some otherwise kinda-evil code.

Even if not: please just write this as a normal conditional inside the loop.

Reply:
I did start out with "always touch", but that really gets in the way. `make` is not good with files that have sub-second resolution time differences, this leads to strange and not always 100% reliably reproducible failures, at least on my system. This does not happen during a "real" build, as building a webpack or npm install take way more than one second. The only alternative that I found was to add `time.sleep(1)` in between, but I found that way more ugly.

Reply:
BTW, the commit message does explain that.

Reply:
I also added a comment right here.